packages/hooks/src/base.ts

```diff
@@ -44,52 +44,49 @@ export class HookManager {
   }
 
   getMiddleware (): Middleware[]|null {
-    const previous = this._parent?.getMiddleware();
+    const previous = this._parent && this._parent.getMiddleware();
 
-    if (previous) {
-      if (this._middleware) {
-        return previous.concat(this._middleware);
-      }
-
-      return previous;
+    if (previous && this._middleware) {
+      return previous.concat(this._middleware);
     }
 
-    return this._middleware;
+    return previous || this._middleware;
   }
 
   collectMiddleware (self: any, _args: any[]): Middleware[] {
     const otherMiddleware = getMiddleware(self);
     const middleware = this.getMiddleware();
 
-    if (otherMiddleware) {
-      if (middleware) {
-        return otherMiddleware.concat(middleware);
-      }
-
-      return otherMiddleware;
+    if (otherMiddleware && middleware) {
+      return otherMiddleware.concat(middleware);
     }
 
-    return middleware;
+    return otherMiddleware || middleware;
   }
 
   props (props: HookContextData) {
     if (!this._props) {
       this._props = {};
     }
 
-    Object.assign(this._props, props);
+    copyProperties(this._props, props);
 
     return this;
   }
 
   getProps (): HookContextData {
-    const previous = this._parent?.getProps();
-
-    if (previous) {
-      return Object.assign({}, previous, this._props);
+    const previous = this._parent && this._parent.getProps();
+
+    if (previous && this._props) {
+      if (this._props) {
+        const result = {};
```

Comment:
I think I get why `copyProperties` is needed here. What I was thinking is that we might want to be able to use it the same way as `Object.assign`.